of_design_doc.md

```diff
@@ -66,21 +66,25 @@ MatrixType = @of(
     data=of(Array, rows, cols),
 )
 
-# Create concrete type by specifying constants
-ConcreteType = MatrixType(;rows=3, cols=4)
-# Can also validate data while concretizing
-ConcreteType = MatrixType(;rows=3, cols=4, data=rand(3, 4))  # Validates that data matches 3x4
-data_nt = (data=rand(3, 4),)
-flat = flatten(ConcreteType, data_nt)
-reconstructed = unflatten(ConcreteType, flat)
+# Create instance by providing all constants (default to zero for data)
+instance = MatrixType(;rows=3, cols=4)  
+# instance = (data = zeros(3, 4),)
 
-SemiConcreteType = MatrixType(; rows=3) # this will return a type with `cols` eliminated
+# Create instance with specific data
+instance = MatrixType(;rows=3, cols=4, data=rand(3, 4))  
+# instance = (data = <provided 3x4 matrix>,)
 
-# rand and zero will first concretize the type and call rand
-rand(MatrixType(; rows=3, cols=4))
-zero(MatrixType(; rows=10, cols=5))
+# Use flatten/unflatten with unconcretized types (providing constants as kwargs)
+flat = flatten(MatrixType, instance; rows=3, cols=4)
+reconstructed = unflatten(MatrixType, flat; rows=3, cols=4)
```

Comment:
We can use `of(MatrixType::OfType; rows=3, cols=4)` for creating semi-concrete types, where `rows=3, cols=4` serves a similar role to `lower` and `upper`. 

```suggestion
reconstructed = unflatten(of(MatrixType; rows=3, cols=4), flat; rows=3, cols=4)
```

Note that the first argument of this syntax has to be a subtype of `OfRype`.

Reply:
yeah, this is already implemented, I can update the design doc